### PR TITLE
feat: Flax

### PR DESCRIPTION
**Pull request by aarnphm**
This PR adds support for Flax. 
- add data container for Jax.

I will add everything here and then cherry-pick out commits later.

Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

### CODE REVIEW

File: src/bentoml/_internal/frameworks/flax.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+            f.write(serialization.to_bytes(state))
+        return bento_model
+
+
+def get_runnable(bento_model: bentoml.Model) -> t.Type[bentoml.Runnable]:
+    """
+    Private API: use :obj:`~bentoml.Model.to_runnable` instead.
+    """
+    partial_kwargs: dict[str, t.Any] = bento_model.info.options.partial_kwargs
+    device: str = bento_model.info.options.device
```
Comment (larme) [inline src/bentoml/_internal/frameworks/flax.py:269]: again why do we choose device using options? If we want user to choose device freely, should we also have a fall back strategy if user does not provide any device?

File: src/bentoml/_internal/frameworks/flax.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        method_partial_kwargs = partial_kwargs.get(method_name)
+        if method_partial_kwargs is not None:
+            method = functools.partial(method, **method_partial_kwargs)
+
+        def mapping(item: jnp.ndarray) -> jnp.ndarray:
+            if not LazyType["jnp.ndarray"]("jnp.ndarray").isinstance(item):
+                return jnp.asarray(item)
+            return item
+
+        def run_method(self: FlaxRunnable, *args: jnp.ndarray):
```
Comment (larme) [inline src/bentoml/_internal/frameworks/flax.py:293]: so the type of  inputs of flax runner can be only `jnp.ndarray`? Should we consider adding support for ndarray? or other tensor format?

**Review by larme — CHANGES_REQUESTED** [review]

File: examples/flax/MNIST/README.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# MNIST classifier
+
+This project demonstrates a simple CNN for MNIST classifier served with BentoML.
+
+### Instruction
+
+Run training scripts:
+
+```bash
+bazel run :train -- --num-epochs 2
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply examples/flax/MNIST/README.md:10]: I align this with #3074. We will probably want to migrate to bazel as our infra.

File: src/bentoml/_internal/frameworks/common/jax.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        cls,
+        batch: jnp.ndarray,
+        batch_dim: int = 0,
+        plasma_db: ext.PlasmaClient | None = Provide[BentoMLContainer.plasma_db],
+    ) -> Payload:
+        if plasma_db:
+            return cls.create_payload(
+                plasma_db.put(batch).binary(),
+                batch.shape[batch_dim],
+                {"plasma": True},
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/common/jax.py:73]: I think we probably have to convert jnp.array to numpy array.

File: src/bentoml/_internal/frameworks/flax.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+            params = Params[jnp.ndarray](*args).map(mapping)
+
+            arg = params.args[0] if len(params.args) == 1 else params.args
+            # NOTE: can we jit this?
+            # No?, as we should not interfere with JAX tracing in multiple threads
+            # https://jax.readthedocs.io/en/latest/concurrency.html?highlight=concurrency
+            return self.model.apply({"params": self.params}, arg, method=method)
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/flax.py:300]: I believe apply already returns multiple outputs.

File: tests/integration/frameworks/models/flax.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    return params
+
+
+def is_close(model: nn.Module, state_dict: dict[str, t.Any], arr: jnp.ndarray):
+    def check(out: jnp.ndarray) -> bool:
+        logit = model.apply({"params": state_dict["params"]}, arr)
+        chex.assert_equal_shape([logit, out])
+        return True
+
+    return check
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply tests/integration/frameworks/models/flax.py:70]: sg.

File: src/bentoml/_internal/frameworks/flax.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+       import bentoml
+       import jax
+
+       net, state_dict = bentoml.flax.load_model("mnist:latest")
+       predict_fn = jax.jit(lambda s: net.apply({"params": state_dict["params"]}, x))
+       results = predict_fn(jnp.ones((1, 28, 28, 1)))
+    """
+    # NOTE: we need to hide all GPU from TensorFlow, otherwise it will try to allocate
+    # memory on the GPU and make it unavailable for JAX.
+    tf.config.experimental.set_visible_devices([], "GPU")
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/flax.py:129]: We have to hide GPU for TensorFlow so that Jax can allocate the DeviceArray to the GPU.

[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/flax.py:129]: This has to do with XLA are still part of tensorflow 😐

File: src/bentoml/_internal/frameworks/flax.py (hunk 2 above)
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/flax.py:293]: I was thinking to only support jnp.ndarray for now. This is because we don't know how to process each data type, and users have to handle that for now.

File: src/bentoml/_internal/frameworks/flax.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (larme) [reply src/bentoml/_internal/frameworks/flax.py:300]: ok it seems return tuple, not sure why I thought it return a list

File: src/bentoml/_internal/frameworks/common/jax.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (larme) [reply src/bentoml/_internal/frameworks/common/jax.py:73]: ye, can this conversion be inplace?

File: src/bentoml/_internal/frameworks/flax.py (hunk 2 above)
Comment (larme) [reply src/bentoml/_internal/frameworks/flax.py:293]: can we support `np.ndarray`?

File: examples/flax/MNIST/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (larme) [reply examples/flax/MNIST/README.md:10]: internally we should use bazel. But for examples/documentations should we assume that user is familiar with bazel? My feeling is that bazel has more mind share in C++ audience than in python user.

[the remark this answers is not in the document]
Comment (aarnphm) [reply examples/flax/MNIST/README.md:10]: the scripts are still there. We can show it doing either in bazel and just in plain script?

[the remark this answers is not in the document]
Comment (larme) [reply examples/flax/MNIST/README.md:10]: ye let's do both

File: examples/flax/MNIST/train.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+) -> train_state.TrainState:
+    """Execute model training and evaluation loop.
+    Args:
+      config: Hyperparameter configuration for training and evaluation.
+      workdir: Directory where the tensorboard summaries are written to.
+    Returns:
+      The train state (which includes the `.params`).
+    """
+    train_ds, test_ds = get_datasets()
+    rng = jax.random.PRNGKey(0)
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply examples/flax/MNIST/train.py:153]: The seed is actually already randomized during training set via jax.random.split.

Still, I will add a `--hermetic` flag to determine whether to use randomize rng for training state (by default hermetic is False)

File: examples/flax/MNIST/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (aarnphm) [reply examples/flax/MNIST/README.md:15]: Done now.

File: src/bentoml/_internal/frameworks/flax.py (hunk 2 above)
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/flax.py:288]: Yes.

**Comment by aarnphm** [comment]
Hi @larme, PTAL.

I'm intending to use this to build some of the example projects using Flax.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -11,6 +11,7 @@ env:
   LINES: 120
   COLUMNS: 120
   BENTOML_DO_NOT_TRACK: True
+  PYTEST_PLUGINS: bentoml.testing.pytest.plugin
```
Comment (sauyon) [inline .github/workflows/ci.yml:14]: let's put this in another PR so we can get this in quickly?

File: docs/source/frameworks/pytorch.rst [hunk 10]
```diff
@@ -21,7 +21,7 @@ For common PyTorch models with single input:
 
 .. code-block:: python
     :caption: `train.py`
-
+
```
Comment (larme) [inline docs/source/frameworks/pytorch.rst:24]: this file has some spaces in blank lines, we should remove them

File: src/bentoml/_internal/frameworks/common/jax.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    def from_payload(
+        cls,
+        payload: Payload,
+        plasma_db: ext.PlasmaClient | None = Provide[BentoMLContainer.plasma_db],
+    ) -> jnp.ndarray:
+        if payload.meta.get("plasma"):
+            import pyarrow.plasma as plasma
+
+            assert plasma_db
+            return plasma_db.get(plasma.ObjectID(payload.data))
```
Comment (larme) [inline src/bentoml/_internal/frameworks/common/jax.py:99]: when we `put` we convert `jnp.array` to `np.array`, so here we should convert it back?

File: src/bentoml/_internal/frameworks/common/jax.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        return pickle.loads(payload.data)
```
Comment (larme) [inline src/bentoml/_internal/frameworks/common/jax.py:100]: same here, should we convert it back?

File: src/bentoml/_internal/frameworks/flax.py [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+            self.params = self.state_dict["params"]
+            self.methods_cache: t.Dict[str, t.Callable[..., t.Any]] = {}
+
+    def gen_run_method(self: FlaxRunnable, method_name: str):
+        method = getattr(self.model, method_name)
+        method_partial_kwargs = partial_kwargs.get(method_name)
+        if method_partial_kwargs is not None:
+            method = functools.partial(method, **method_partial_kwargs)
+
+        def mapping(item: jnp.ndarray) -> jnp.ndarray:
```
Comment (larme) [inline src/bentoml/_internal/frameworks/flax.py:281]: the type of `item` should be `jnp.ndarray | np.ndarray | pandas.DataFrame` if we are doing the conversion ourselves

File: src/bentoml/_internal/frameworks/flax.py [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+            if LazyType["ext.NpNDArray"]("numpy.ndarray").isinstance(item):
+                return jnp.asarray(item)
+            if LazyType["ext.PdDataFrame"]("pandas.DataFrame").isinstance(item):
+                return jnp.asarray(item.to_numpy())
+            return item
+
+        def run_method(self: FlaxRunnable, *args: jnp.ndarray):
+            params = Params[jnp.ndarray](*args).map(mapping)
```
Comment (larme) [inline src/bentoml/_internal/frameworks/flax.py:289]: here the type also should be `jnp.ndarray | np.ndarray | pandas.DataFrame`

File: src/bentoml/_internal/frameworks/flax.py [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+            f.write(serialization.to_bytes(state))
+        return bento_model
+
+
+def get_runnable(bento_model: bentoml.Model) -> t.Type[bentoml.Runnable]:
+    """Private API: use :obj:`~bentoml.Model.to_runnable` instead."""
+    partial_kwargs: dict[str, t.Any] = bento_model.info.options.partial_kwargs
+
+    class FlaxRunnable(bentoml.Runnable):
+        SUPPORTED_RESOURCES = ("nvidia.com/gpu", "cpu", "tpu")
```
Comment (larme) [inline src/bentoml/_internal/frameworks/flax.py:264]: Should we put "tpu" at the first?

File: tests/integration/frameworks/models/flax.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+def assert_equal_shape(
+    model: nn.Module, state_dict: dict[str, t.Any], arr: jnp.ndarray
+):
+    def check(out: jnp.ndarray) -> bool:
+        logit = model.apply({"params": state_dict["params"]}, arr)
+        chex.assert_equal_shape([logit, out])
+        return True
+
+    return check
```
Comment (larme) [inline tests/integration/frameworks/models/flax.py:71]: last time I forgot to ask, why we cannot do a value equality test here? Our model should output the same (or very close) value as our runner?

**Review by larme — CHANGES_REQUESTED** [review]
I think we may need minor changes here but otherwise everything looks quite good now

File: docs/source/frameworks/pytorch.rst (hunk 10 above)
Comment (aarnphm) [reply docs/source/frameworks/pytorch.rst:24]: got it.

File: src/bentoml/_internal/frameworks/common/jax.py (hunk 11 above)
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/common/jax.py:99]: I'm thinking to just remove this section, considering PlasmaClient will be deprecated soon.

File: tests/integration/frameworks/models/flax.py (hunk 16 above)
Comment (aarnphm) [reply tests/integration/frameworks/models/flax.py:71]: oh I haven't figured out why I can't seem to check is_close between two results.

File: src/bentoml/_internal/frameworks/common/jax.py (hunk 11 above)
Comment (larme) [reply src/bentoml/_internal/frameworks/common/jax.py:99]: I think we can do that

File: src/bentoml/_internal/frameworks/common/jax.py (hunk 12 above)
Comment (aarnphm) [reply src/bentoml/_internal/frameworks/common/jax.py]: addressed.

File: tests/integration/frameworks/models/flax.py (hunk 16 above)
Comment (aarnphm) [reply tests/integration/frameworks/models/flax.py]: addressed.

**Comment by aarnphm** [comment]
Will follow up with a example PR for flax and frameworks docs afterward.

**Comment by sauyon** [comment]
Remind me to take a look at this tomorrow if Larme doesn't get a chance to.

**Review by larme — APPROVED** [review]
👍LGTM

Let's add an equality test in a later PR

**Comment by aarnphm** [comment]
> 👍LGTM
> 
> Let's add an equality test in a later PR

I have already added that.

File: pyproject.toml [hunk 17]
```diff
@@ -236,7 +236,7 @@ exclude = '''
 '''
 
 [tool.pytest.ini_options]
-addopts = ["-rfEX", "-pbentoml.testing.pytest.plugin"]
```
Comment (sauyon) [inline pyproject.toml:239]: @aarnphm why was this in this PR anyway?

Comment (aarnphm) [reply pyproject.toml:239]: Oh because the test in the examples will always running train everytime the test is running.

THe plugins should be for our internal test only.

Comment (aarnphm) [reply pyproject.toml:239]: We can revert this back imo.

Comment (sauyon) [reply pyproject.toml:239]: Why should it be for our internal tests only?

Comment (sauyon) [reply pyproject.toml:239]: Given we have implemented as part of our library, it seems to me like it should be general-purpose?

Comment (aarnphm) [reply pyproject.toml:239]: i will put up a PR to revert this.
